internal/luks2/metadata.go

```diff
@@ -819,56 +819,73 @@ func ReadHeader(path string, lockMode LockMode) (*HeaderInfo, error) {
 	defer f.Close()
 
 	// Try to decode and check the primary header
-	primaryHdr, primaryJSON, primaryErr := decodeAndCheckHeader(f, 0, true)
+	primaryHdr, primaryJSONData, primaryErr := decodeAndCheckHeader(f, 0, true)
+	var primaryMetadata Metadata
+	if primaryErr == nil {
+		if err := json.NewDecoder(primaryJSONData).Decode(&primaryMetadata); err != nil {
+			primaryErr = xerrors.Errorf("cannot decode JSON metadata area: %w", err)
+		}
+	}
 
 	var secondaryHdr *binaryHdr
-	var secondaryJSON *bytes.Buffer
+	var secondaryJSONData *bytes.Buffer
+	var secondaryErr error
 	if primaryErr != nil {
 		// No valid primary header. Try to decode and check a secondary header from one of the
 		// well known offsets.
 		for _, off := range []int64{0x4000, 0x8000, 0x10000, 0x20000, 0x40000, 0x80000, 0x100000, 0x200000, 0x400000} {
-			secondaryHdr, secondaryJSON, err = decodeAndCheckHeader(f, off, false)
-			if err == nil {
+			secondaryHdr, secondaryJSONData, secondaryErr = decodeAndCheckHeader(f, off, false)
+			if secondaryErr == nil {
 				break
 			}
 		}
 	} else {
 		// Try to decode and check the secondary header immediately after the primary header.
-		secondaryHdr, secondaryJSON, _ = decodeAndCheckHeader(f, int64(primaryHdr.HdrSize), false)
+		secondaryHdr, secondaryJSONData, secondaryErr = decodeAndCheckHeader(f, int64(primaryHdr.HdrSize), false)
```

Comment:
my question was because  in if true stanza starts at 0x4000, this one uses int64(primaryHdr.HdrSize) so I'm left wondering if they are the same? or we skip the first possible position for the secondary if the primary can't be read or yet something else?